buildpacktoml.go

```diff
@@ -0,0 +1,336 @@
+		Build bool `toml:"build"`
+	}
+	_, err := toml.DecodeFile(path, &layerTOML)
+	return err == nil && layerTOML.Build
+}
+
+func (b *BuildpackTOML) readOutputFiles(bpLayersDir, bpPlanPath string, bpPlanIn BuildpackPlan) (BuildResult, error) {
+	br := BuildResult{}
+	bpFromBpInfo := GroupBuildpack{ID: b.Buildpack.ID, Version: b.Buildpack.Version}
+
+	// setup launch.toml
+	var launchTOML LaunchTOML
+	launchPath := filepath.Join(bpLayersDir, "launch.toml")
+
+	if api.MustParse(b.API).Compare(api.MustParse("0.5")) < 0 { // buildpack API <= 0.4
+		// read buildpack plan
+		var bpPlanOut BuildpackPlan
+		if _, err := toml.DecodeFile(bpPlanPath, &bpPlanOut); err != nil {
+			return BuildResult{}, err
+		}
+
+		// set BOM and MetRequires
+		if err := validateBOM(bpPlanOut.toBOM(), b.API); err != nil {
+			return BuildResult{}, err
+		}
+		br.BOM = withBuildpack(bpFromBpInfo, bpPlanOut.toBOM())
+		br.MetRequires = names(bpPlanOut.Entries)
+
+		// read launch.toml, return if not exists
+		if _, err := toml.DecodeFile(launchPath, &launchTOML); os.IsNotExist(err) {
+			return br, nil
+		} else if err != nil {
+			return BuildResult{}, err
+		}
+	} else {
+		// read build.toml
+		var bpBuild BuildTOML
+		buildPath := filepath.Join(bpLayersDir, "build.toml")
+		if _, err := toml.DecodeFile(buildPath, &bpBuild); err != nil && !os.IsNotExist(err) {
+			return BuildResult{}, err
+		}
+		if err := validateBOM(bpBuild.BOM, b.API); err != nil {
+			return BuildResult{}, err
+		}
+
+		// set MetRequires
+		if err := validateUnmet(bpBuild.Unmet, bpPlanIn); err != nil {
+			return BuildResult{}, err
+		}
+		br.MetRequires = names(bpPlanIn.filter(bpBuild.Unmet).Entries)
+
+		// read launch.toml, return if not exists
+		if _, err := toml.DecodeFile(launchPath, &launchTOML); os.IsNotExist(err) {
+			return br, nil
+		} else if err != nil {
+			return BuildResult{}, err
+		}
+
+		// set BOM
+		if err := validateBOM(launchTOML.BOM, b.API); err != nil {
+			return BuildResult{}, err
+		}
+		br.BOM = withBuildpack(bpFromBpInfo, launchTOML.BOM)
+	}
+
+	// set data from launch.toml
+	br.Labels = append([]Label{}, launchTOML.Labels...)
+	for i := range launchTOML.Processes {
+		launchTOML.Processes[i].BuildpackID = b.Buildpack.ID
+	}
+	br.Processes = append([]launch.Process{}, launchTOML.Processes...)
+	br.Slices = append([]layers.Slice{}, launchTOML.Slices...)
+
+	return br, nil
+}
+
+func validateBOM(bom []BOMEntry, bpAPI string) error {
+	if api.MustParse(bpAPI).Compare(api.MustParse("0.5")) < 0 {
+		for _, entry := range bom {
+			if version, ok := entry.Metadata["version"]; ok {
+				metadataVersion := fmt.Sprintf("%v", version)
+				if entry.Version != "" && entry.Version != metadataVersion {
+					return errors.New("top level version does not match metadata version")
+				}
+			}
+		}
+	} else {
+		for _, entry := range bom {
+			if entry.Version != "" {
+				return fmt.Errorf("bom entry '%s' has a top level version which is deprecated", entry.Name)
```

Comment:
This is an error so we shouldn't call it "deprecated".
```suggestion
				return fmt.Errorf("bom entry '%s' has a top level version which is not allowed. The buildpack should instead set metadata.version", entry.Name)
```